src/Stratis.Bitcoin.Features.SmartContracts/Consensus/Rules/AllowedScriptTypesRule.cs

```diff
@@ -0,0 +1,93 @@
+using Stratis.Bitcoin.Consensus.Rules;
+using Stratis.Bitcoin.Features.MemoryPool;
+using Stratis.SmartContracts;
+using Stratis.SmartContracts.Core;
+using Block = NBitcoin.Block;
+
+namespace Stratis.Bitcoin.Features.SmartContracts.Consensus.Rules
+{
+    /// <summary>
+    /// Enforces that only certain script types are used on the network.
+    /// </summary>
+    public class AllowedScriptTypesRule : PartialValidationConsensusRule, ISmartContractMempoolRule
+    {
+        protected ISmartContractCoinviewRule ContractCoinviewRule { get; private set; }
+
+        /// <inheritdoc />
+        public override void Initialize()
+        {
+            base.Initialize();
+            this.ContractCoinviewRule = (ISmartContractCoinviewRule)this.Parent;
+        }
+
+        public override Task RunAsync(RuleContext context)
+        {
+            Block block = context.ValidationContext.BlockToValidate;
+
+            foreach (Transaction transaction in block.Transactions)
+            {
+                CheckTransaction(transaction);
+            }
+
+            return Task.CompletedTask;
+        }
+
+        public void CheckTransaction(MempoolValidationContext context)
+        {
+            CheckTransaction(context.Transaction);
+        }
+
+        private void CheckTransaction(Transaction transaction)
+        {
+            // Why dodge coinbase?
+            // 1) Coinbase can only be written by Authority nodes anyhow.
+            // 2) Coinbase inputs look weird, are tough to validate.
+            if (!transaction.IsCoinBase)
+            {
+                foreach (TxOut output in transaction.Outputs)
+                {
+                    CheckOutput(output);
+                }
+
+                foreach (TxIn input in transaction.Inputs)
+                {
+                    CheckInput(input);
+                }
+            }
+        }
+
+        private void CheckOutput(TxOut output)
+        {
+            if (output.ScriptPubKey.IsSmartContractExec())
+                return;
+
+            if (output.ScriptPubKey.IsSmartContractInternalCall())
+                return;
+
+            if (PayToPubkeyHashTemplate.Instance.CheckScriptPubKey(output.ScriptPubKey))
+                return;
+
+            if (PayToScriptHashTemplate.Instance.CheckScriptPubKey(output.ScriptPubKey))
+                return;
+
+            new ConsensusError("disallowed-output-script", "Only P2PKH, multisig and smart contract scripts are allowed.").Throw();
+        }
+
+        private void CheckInput(TxIn input)
+        {
+            if (input.ScriptSig.IsSmartContractSpend())
+                return;
+
+            if (PayToPubkeyHashTemplate.Instance.CheckScriptSig(this.ContractCoinviewRule.Network, input.ScriptSig))
+                return;
+
+            if (PayToScriptHashTemplate.Instance.CheckScriptSig(this.ContractCoinviewRule.Network, input.ScriptSig, null)) // TODO: This last parameter is a ScriptPubKey. Which ScriptPubKey??
```

Comment:
I would expect (without looking too deeply into the underlying code, admittedly) that this is the P2SH redeemscript's corresponding scriptPubKey. It is not really possible to validate a scriptSig without the corresponding scriptPubKey it was originally created against.

In the case of a P2SH transaction the redeemscript would appear in the scriptSig, and the validation would verify that what was provided hashes/signs correctly.

Reply:
And my comment for `CheckOutput` applies here too, obviously.